### PR TITLE
WIP: Tor support

### PR DESCRIPTION
**Pull request by freehodl**
This is my first ever pull request so please let me know what I'm doing wrong. 

This change adds a button to the node launcher GUI that:
1. Downloads Tor
2. Installs Tor
3. Adds a pre-configured torrc file
4. Writes Tor configuration lines to lnd.conf and bitcoin.conf 

This change only works on Windows so far.

I'm not a Tor expert and so security review is still needed.

### CODE REVIEW

**Comment by PierreRochard** [comment]
Excellent! Thank you!

File: node_launcher/node_set/setup_tor.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+import subprocess
+import zipfile
+import os.path
+from urllib.request import urlopen
+
+from node_launcher.constants import BITCOIN_DATA_PATH, BITCOIN_CONF_PATH, TOR_DATA_PATH, \
+    TOR_TORRC_PATH, LND_CONF_PATH , TOR_PATH, TOR_EXE_PATH, OPERATING_SYSTEM, IS_WINDOWS
+
+
+class setup_tor(object):
```
Comment (PierreRochard) [inline node_launcher/node_set/setup_tor.py]:
```suggestion
class SetupTor(object):
```

File: node_launcher/node_set/node_set.py [hunk 2]
```diff
@@ -32,6 +34,7 @@ def __init__(self, network: Network):
             bitcoin=self.bitcoin
         )
         self.lnd_client = LndClient(self.lnd)
+        self.setup_tor = setup_tor()
```
Comment (PierreRochard) [inline node_launcher/node_set/node_set.py]: Pass in the NodeSet here so you have access to the configuration files

```suggestion
        self.setup_tor = SetupTor(self)
```

File: node_launcher/node_set/setup_tor.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
```
Comment (PierreRochard) [inline node_launcher/node_set/setup_tor.py]:
```suggestion
    def __init__(self, node_set: NodeSet):
        self.node_set = node_set
```

File: node_launcher/node_set/setup_tor.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    def edit_bitcoin_conf(self):
+        f = open(str(BITCOIN_CONF_PATH[OPERATING_SYSTEM]) , 'a')
+        f.write('proxy=127.0.0.1:9050\n')
```
Comment (PierreRochard) [inline node_launcher/node_set/setup_tor.py]:
```suggestion
        self.node_set.bitcoin.file['proxy'] = '127.0.0.1:9050'
```

Comment (PierreRochard) [reply node_launcher/node_set/setup_tor.py]: Same for the rest of Bitcoin/LND conf params

File: node_launcher/node_set/setup_tor.py [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+from node_launcher.constants import BITCOIN_DATA_PATH, BITCOIN_CONF_PATH, TOR_DATA_PATH, \
+    TOR_TORRC_PATH, LND_CONF_PATH , TOR_PATH, TOR_EXE_PATH, OPERATING_SYSTEM, IS_WINDOWS
+
+class SetupTor(object):
+    def __init__(self, node_set):
+        self.node_set = node_set
+
+    def edit_bitcoin_conf(self):
+        f = open(str(BITCOIN_CONF_PATH[OPERATING_SYSTEM]) , 'a')
```
Comment (PierreRochard) [inline node_launcher/node_set/setup_tor.py]: No longer need `f`

File: node_launcher/node_set/setup_tor.py [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        self.node_set.bitcoin.file['proxy'] = '127.0.0.1:9050'
+        self.node_set.bitcoin.file['listen'] = '1'
```
Comment (PierreRochard) [inline node_launcher/node_set/setup_tor.py]:
```suggestion
        self.node_set.bitcoin.file['listen'] = True
```

File: node_launcher/node_set/setup_tor.py [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+        self.node_set.bitcoin.file['bind'] = '127.0.0.1'
+        self.node_set.bitcoin.file['debug'] ='tor'
+        f.close()
+
+    def edit_lnd_conf(self):
+        f = open(str(LND_CONF_PATH[OPERATING_SYSTEM]), 'a')
+        self.node_set.lnd.file['listen'] = 'localhost'
+        self.node_set.lnd.file[' ']
```
Comment (PierreRochard) [inline node_launcher/node_set/setup_tor.py]: Not needed

File: node_launcher/node_set/setup_tor.py [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+        self.node_set.lnd.file['tor.active'] ='1'
```
Comment (PierreRochard) [inline node_launcher/node_set/setup_tor.py]: Same for the rest 

```suggestion
        self.node_set.lnd.file['tor.active'] = True
```

**Comment by PierreRochard** [comment]
Excellent progress @freehodl! I'll get it working on macos and linux and will contribute those pieces asap

**Comment by freehodl** [comment]
Thanks @PierreRochard ! I'll get these changes made asap. I've already started work on the macos piece so I can take care of that if you want to focus on linux?

**Comment by willcl-ark** [comment]
> I love where this is going! It'll be safer to write to the configuration files using the ConfigurationFile class https://github.com/PierreRochard/node-launcher/blob/master/node_launcher/services/configuration_file.py

Just as a point of interest more than anything else at this stage, you might like to check out Core's plan to split the configuration files in two; a read-write one which the GUI can effectively modify (bitcoin-conf-rw) and one which the daemon/GUI won't be able to modify (bitcoin.conf):

https://github.com/bitcoin/bitcoin/pull/12833

It would only require minimal changes for this project though I think as we are intentionally modifying those files (settings) on the user's behalf.

**Comment by PierreRochard** [comment]
Great catch @willcl-ark, I'll keep a close eye on that, doesn't look like it'll be a big change on our end

**Comment by PierreRochard** [comment]
Thanks @freehodl! I'll try this out on Linux this weekend

**Comment by PierreRochard** [comment]
Thank you @freehodl! I'm finally ready to rebase, review, and merge in tor support! I'll loop in (can we still use that phrase!?) @vindard as he has been building on top of your commits

**Comment by PierreRochard** [comment]
I'll rebase over here: https://github.com/lightning-power-users/node-launcher/pull/298

**Comment by PierreRochard** [comment]
Thank you @freehodl  and @vindard ! Sorry it took me so long to get back to this!

**Pull request closed without merging**